In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchsummary import summary
from tqdm import tqdm

from model2 import Net
from train_test_loop import *
from data_utils import *

/home/raghu/.local/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(


In [2]:
torch.manual_seed(1)

In [3]:
BATCH_SIZE=128
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
       BatchNorm2d-3            [-1, 8, 26, 26]              16
           Dropout-4            [-1, 8, 26, 26]               0
            Conv2d-5           [-1, 16, 24, 24]           1,152
              ReLU-6           [-1, 16, 24, 24]               0
       BatchNorm2d-7           [-1, 16, 24, 24]              32
           Dropout-8           [-1, 16, 24, 24]               0
            Conv2d-9           [-1, 10, 24, 24]             160
        MaxPool2d-10           [-1, 10, 12, 12]               0
           Conv2d-11           [-1, 16, 10, 10]           1,440
             ReLU-12           [-1, 16, 10, 10]               0
      BatchNorm2d-13           [-1, 16, 10, 10]              32
          Dropout-14           [-1, 16,

In [4]:
default_train_transforms = transforms.Compose([transforms.RandomRotation((-7.0, 7.0), fill=(1,)),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))])
default_test_transforms = transforms.Compose([transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))])

In [5]:
train_dataloader, test_dataloader = get_augmented_MNIST_dataset("~/work/data/", 
                                    train_tfms=default_train_transforms, test_tfms=default_test_transforms,
                                    batch_sz=BATCH_SIZE)

In [6]:
optimizer = torch.optim.SGD(model.parameters(), lr=0.05, momentum=0.9)
scheduler = get_schedulder(optimizer, "StepLR", step_size=6, gamma=0.25)

In [7]:
EPOCHS=15
for i in range(EPOCHS):
    print("Running EPOCH: "+str(i+1))
    train(model, device, train_dataloader, optimizer=optimizer, epoch=i)
    test(model, device, test_dataloader)

Running EPOCH: 1


Loss=0.025125885382294655 Batch_id=468 Accuracy=91.36: 100%|█████████████████████████| 469/469 [00:01<00:00, 236.28it/s]



Test set: Average loss: 0.0693, Accuracy: 9791/10000 (97.91%)

Running EPOCH: 2


Loss=0.03183881565928459 Batch_id=468 Accuracy=97.69: 100%|██████████████████████████| 469/469 [00:01<00:00, 238.76it/s]



Test set: Average loss: 0.0442, Accuracy: 9866/10000 (98.66%)

Running EPOCH: 3


Loss=0.06838590651750565 Batch_id=468 Accuracy=98.17: 100%|██████████████████████████| 469/469 [00:01<00:00, 245.80it/s]



Test set: Average loss: 0.0451, Accuracy: 9853/10000 (98.53%)

Running EPOCH: 4


Loss=0.01746177114546299 Batch_id=468 Accuracy=98.28: 100%|██████████████████████████| 469/469 [00:01<00:00, 248.66it/s]



Test set: Average loss: 0.0371, Accuracy: 9886/10000 (98.86%)

Running EPOCH: 5


Loss=0.026944464072585106 Batch_id=468 Accuracy=98.38: 100%|█████████████████████████| 469/469 [00:01<00:00, 250.80it/s]



Test set: Average loss: 0.0318, Accuracy: 9907/10000 (99.07%)

Running EPOCH: 6


Loss=0.040501270443201065 Batch_id=468 Accuracy=98.53: 100%|█████████████████████████| 469/469 [00:01<00:00, 248.38it/s]



Test set: Average loss: 0.0249, Accuracy: 9915/10000 (99.15%)

Running EPOCH: 7


Loss=0.042971160262823105 Batch_id=468 Accuracy=98.57: 100%|█████████████████████████| 469/469 [00:01<00:00, 244.00it/s]



Test set: Average loss: 0.0263, Accuracy: 9921/10000 (99.21%)

Running EPOCH: 8


Loss=0.05864468216896057 Batch_id=468 Accuracy=98.73: 100%|██████████████████████████| 469/469 [00:01<00:00, 239.26it/s]



Test set: Average loss: 0.0273, Accuracy: 9906/10000 (99.06%)

Running EPOCH: 9


Loss=0.018115108832716942 Batch_id=468 Accuracy=98.70: 100%|█████████████████████████| 469/469 [00:01<00:00, 247.84it/s]



Test set: Average loss: 0.0268, Accuracy: 9919/10000 (99.19%)

Running EPOCH: 10


Loss=0.025244640186429024 Batch_id=468 Accuracy=98.73: 100%|█████████████████████████| 469/469 [00:01<00:00, 244.48it/s]



Test set: Average loss: 0.0251, Accuracy: 9911/10000 (99.11%)

Running EPOCH: 11


Loss=0.05009814724326134 Batch_id=468 Accuracy=98.75: 100%|██████████████████████████| 469/469 [00:01<00:00, 245.38it/s]



Test set: Average loss: 0.0254, Accuracy: 9917/10000 (99.17%)

Running EPOCH: 12


Loss=0.05259786918759346 Batch_id=468 Accuracy=98.73: 100%|██████████████████████████| 469/469 [00:01<00:00, 238.08it/s]



Test set: Average loss: 0.0251, Accuracy: 9918/10000 (99.18%)

Running EPOCH: 13


Loss=0.06490005552768707 Batch_id=468 Accuracy=98.87: 100%|██████████████████████████| 469/469 [00:01<00:00, 247.31it/s]



Test set: Average loss: 0.0246, Accuracy: 9918/10000 (99.18%)

Running EPOCH: 14


Loss=0.011960762552917004 Batch_id=468 Accuracy=98.86: 100%|█████████████████████████| 469/469 [00:02<00:00, 231.12it/s]



Test set: Average loss: 0.0232, Accuracy: 9926/10000 (99.26%)

Running EPOCH: 15


Loss=0.013371418230235577 Batch_id=468 Accuracy=98.86: 100%|█████████████████████████| 469/469 [00:01<00:00, 243.20it/s]



Test set: Average loss: 0.0227, Accuracy: 9927/10000 (99.27%)



#### Target:

Reduce the model size to less than 8k params (primary goal), and tweak with lr/scheduler

#### Result:

Accuracy nearly 99.2% but fluctuating. Model size: 7736 params


#### Analysis:

Now trying to get the model size down first. The change is in the backbone CNN which goes like this: input block -> 2 conv blocks (conv + ReLU) -> transition block (2x2 maxpool) -> 4 conv blocks (conv + ReLU) -> final conv2d layer. The difference from Step 1 is the number of channels: Step1 has a channel progression like 1 -> 32 -> 64 -> 128 -> 32 -> 64 -> 128 -> 10 -> 10, where as we don't cross 16 channels in any conv2d layer here. The new channel progression is 1 -> 8 -> 16 -> 10 -> 16 -> 16 -> 10 with a GAP of kernel size 6 before the last layer. CHange the lr scheduler with lr = 0.05, step_size = 6 and gamma = 0.25. The final change is the removal of one conv block just before the GAP layer, as an experiment to reduce model size, and the number of params is 7,736 (within our range). I selected that particular conv block since the number of input and output channels are the same, so I thought I could remove since number of input and output features is the same. (May work for smaller datasets like MNIST, but may fail on larger datasets since many more features are present in larger datasets). The test accuracy for 15 epochs: 99.2% but is fluctuating. It is closer to the required test accuracy, it is not consistent, so may be we are reducing the lr too fast (i.e. we should increase step_size or decrease gamma or increase lr or a combination of these three).